### DISCUSSION TITLE
Is there any way to ensure that input images will be at least as large as certain dimensions?

### DISCUSSION THREAD

**Discussion by jmlipman** [TorchIO-project/torchio#786] [Q&A]
Consider a heterogeneous dataset comprised of images whose size ranges from 75x190x190 to 300x250x250 and you want to train a CNN on patches of 80x200x200. Is it possible to do this with the existing data transformation functions?

I guess that CropOrPad((80, 200, 200)) won't help because it will crop larger images, and I still want these images to be large so that different patches can be extracted from them. I also guess that Pad(()) won't help either because it will pad all images similarly.

Would a MaybePad(minimum_size=(80, 200, 200)) function that only pads images smaller than the given size be a solution, or is there any way to solve this with the existing functions?

**Comment by romainVala** [comment]
Yes I thing, `CropOrPad` transform is what you search with input `target_shape`

**Reply by jmlipman** [reply]
But if I CropOrPad((80, 200, 200)) an image of size 300x250x250 I will get a 80x200x200 image (patch), and the Queue class won't be able to extract patches from it, so the network won't "see" the whole image. I noticed that there is a function that does something similar to what I described but not exactly the same: EnsureShapeMultiple.

**Comment by romainVala** [comment]
why then not choose the largest image size : 
CropOrPad([300x250x250]) ?

the problem, is that for 'small images' you may extract a patch with full zeros ... (you could still play with WeightedSampler if you have a mask of the original image, in order to restrict patch center within the mask) 

Why are yours images so heterogeneous Is it the field of view (FOV) that is changing or also the voxel size? 
if the voxel size, is changing, you should first resample all image to a given voxel size (and then CropOrPad to account for different FOV)

**Reply by jmlipman** [reply]
As you've said, the problem of using CropOrPad in that way is that patches will be superbig and full of zeros.

Voxel size is the same (since I've resampled all images). The main difference is the field of view, i.e., some images have many more slices than others.

I don't think it's too difficult to write a MaybePad transform and I will probably end up doing that, but maybe I was missing other simpler way to do this.

**Comment by romainVala** [comment]
I do not understand what you have in mind with MaybePad ... to pad only small images ? (this is what CorpOrPad already doing)

Actually I wonder, if you really need, to CropOrPad, if you are using the Queue, (to extract different patch from each volume) then you will enp up with a fixe patch size (whatever the original image size is) ... Though I am not sure if the Queue can handle volume with different size ... (to test) ...

**Reply by romainVala** [reply]
ok sorry, I re read you first post, and I understand, 
you want to pad 'small images' to a minimal size but not crop the bigger one ...
so that the queue with extract patch with image information (so the queue is working fine with image of different size right?)

**Comment by romainVala** [comment]
So may be the best option is  indeed with a MaybePad, (it should be easy, you can do the same as CropOrPad, but with Pad only ... Or just add an extra argument to skip the Croping part in the CropOrPad transform,   It will then pad to your target size, if needed (but not crop) ...

Other option is to construct an appropriate mask, to be use in the WeightedSampler to restrict the patch center of the extraction ...

**Reply by jmlipman** [reply]
Yeah, I ended up creating a MaybePad (it only took 10 lines) and it seems to work well. I haven't thought about the WeightedSampler : )

**Comment by romainVala** [comment]
may be it worth a PR (@fepegar  ?)
not sure what the best, beween to add an argument to CropOrPad or to create a new transform

**Comment by fepegar** [comment]
Hi, both. What about adding an option in the sampler to pad the volume to match the patch size if the patch is larger than the volume?

**Comment by romainVala** [comment]
yes make sense, it will solve the issue I think
